### PR TITLE
str::lines and BufRead::lines: document trailing bare cr behaviour

### PR DESCRIPTION
**Pull request by ijackson**
Apropos discussion here  https://github.com/rust-lang/rust/pull/91047#issuecomment-974321038

Sadly, str::lines gets this wrong.  I think it is probably too late to
fix this, so document it instead.

### CODE REVIEW

File: library/core/src/str/mod.rs [hunk 1]
```diff
@@ -899,6 +899,19 @@ impl str {
...
+    ///
+    /// Handling of a trailing bare CR is, sadly, anomalous:
+    /// (`std::io::BufRead::lines` handles this case correctly.)
+    ///
+    /// ```
+    /// let text = "foo\nbar\r";
+    /// let mut lines = text.lines();
+    ///
+    /// assert_eq!(Some("foo"), lines.next());
+    /// assert_eq!(Some("bar"), lines.next()); // should really return "bar\r"
```
Comment (r00ster91) [inline library/core/src/str/mod.rs:911]:
```suggestion
    /// assert_eq!(Some("bar"), lines.next()); // does not return "bar\r"
```

File: library/std/src/io/mod.rs [hunk 2]
```diff
@@ -2281,6 +2281,19 @@ pub trait BufRead: Read {
     /// assert_eq!(lines_iter.next(), None);
     /// ```
     ///
+    /// Unlike [`str::lines`], trailing bare CR is handled correctly:
```
Comment (r00ster91) [inline library/std/src/io/mod.rs:2284]:
```suggestion
    /// Unlike [`str::lines`], trailing bare CR is not handled as a newline:
```

File: library/core/src/str/mod.rs [hunk 3]
```diff
@@ -899,6 +899,19 @@ impl str {
     ///
     /// assert_eq!(None, lines.next());
     /// ```
+    ///
+    /// Handling of a trailing bare CR is, sadly, anomalous:
+    /// (`std::io::BufRead::lines` handles this case correctly.)
```
Comment (r00ster91) [inline library/core/src/str/mod.rs:904]:
```suggestion
    /// Unlike [`std::io::BufRead::lines`], trailing bare CR is handled as a newline:
```

File: library/std/src/io/mod.rs [hunk 4]
```diff
@@ -2281,6 +2281,19 @@ pub trait BufRead: Read {
...
+    ///
+    /// ```
+    /// use std::io::{self, BufRead};
+    ///
+    /// let cursor = io::Cursor::new(b"lorem\nipsum\r");
+    ///
+    /// let mut lines_iter = cursor.lines().map(|l| l.unwrap());
+    /// assert_eq!(lines_iter.next(), Some(String::from("lorem")));
+    /// assert_eq!(lines_iter.next(), Some(String::from("ipsum\r"))); // bare CR is not a newline
```
Comment (r00ster91) [inline library/std/src/io/mod.rs:2293]:
```suggestion
    /// assert_eq!(lines_iter.next(), Some(String::from("ipsum\r"))); // does not return "ipsum"
```

**Review by ghost — CHANGES_REQUESTED** [review]
Can we keep it neutral and not have it sound so negative? I don't think giving the "we messed up" vibe in the official docs adds a lot of value.
And instead of referring to the two ways as "correct" and "incorrect" you could be more explicit.

**Comment by hkratz** [comment]
I think we should just fix this and make `str::lines` like `BufRead::lines`. I don't think that much (if any) code relies on this (faulty) behavior.

**Comment by ijackson** [comment]
> I think we should just fix this and make `str::lines` like `BufRead::lines`. I don't think that much (if any) code relies on this (faulty) behavior.

I think this would need an FCP, but it is indeed our other option.  I will work up an MR to do this, so we can see what it looks like (and what kind of justification I find I am able to write).

I think we should choose between fixing it, and acknowledging that it is, at the very least, anomalous.  I don't think we should pretend (even implicitly) in the docs that we think the current behaviour of `str::lines` is correct.

**Comment by shirshak55** [comment]
I don't think many libs rely on this behaviour so fixing would be a wise decision rather than mitigating it by adding docs. Not many of us are going to read the documentation and this is not expected too.  So it would be better if we fix it.  Maybe next time when we do a crater run we can investigate how much churn it would cause in the ecosystem.

**Comment by Dylan-DPC** [comment]
@ijackson if you can rebase this we can push this forward

**Comment by ChrisDenton** [comment]
> I think we should just fix this and make `str::lines` like `BufRead::lines`.

> I don't think many libs rely on this behaviour so fixing would be a wise decision rather than mitigating it by adding docs.

Given these comments, ping @rust-lang/libs-api

**Comment by joshtriplett** [comment]
I do think we could try a crater run fixing the two to be consistent.

**Comment by m-ou-se** [comment]
I agree, these two being inconsistent seems like a mistake we should fix.

**Comment by JohnCSimon** [comment]
Triage: 
@ijackson What is the status of this PR?

**Comment by JohnCSimon** [comment]
@ijackson 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**